doc/compiler/osr/OSR.md

```diff
@@ -10,36 +10,241 @@ is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+**Q**: seems like a circular dependence between TR_OSRPoint and TR_OSRMethodData: the latter describes mutiple points but then is associated with one point?
+
+Array of **TR_OSRMethodData**: has an entry for each inlined site index in the method being compiled. Eventually, it is used to populate **Instr2SharedSlotMetaData**.
+
+## Runtime
+
+**OSRBuffer**: the buffer that is populated by JITed code with locals and operands for all active frames. It is used by VM to reconstruct the frames. **OSRBuffer** is created by the VM. Since the VM is aware of the number of auto slots, parameter slots,
+and the maximum number of pending push temporary slots, it can always ensure the buffer is large enough to hold all autos,
+parameters and pending push temporaries for all interpreter frames the VM needs to recreate at a particular OSR point.
+
+**OSRScratchBuffer**: the buffer is populated by JITed code with symbols that share slots. It is used by **prepareForOSRHelper** together with **Instr2SharedSlotMetaData** to copy the right symbols into **OSRBuffer**.
+
+**Instr2SharedSlotMetaData**: each entry contains an instruction offset of an OSR induction point within the compiled method. It is followed by an array of tuples. Each tuple contains inined site index and indicates from which offset in **OSRScracthBuffer** to which offset in **OSRBuffer** the symbol needs to be copied by the corresponding **prepareForOsrHelper**. Therefore, given an offset of an OSR induction point, this metadata describes which JIT symbols need to be copied into **OSRBuffer** for each frame.
+
+**prepareForOsrHelper**: a helper call that is inserted in the IL with the purpose of keeping all necessary symbol live at a transition point. It also "logically" copies JIT symbols into **OSRBuffer**. Although, during trees lowering, symbols are explicitly copied in the IL, except for the ones that share slots. Then, the helper is only used for copying symbols that require slot sharing info. **prepareForOsrHelper** takes its inlined site index as a parameter.
+
+
+# IL Generation
+
+During IL generation, we examine each bytecode to see if it is an OSR yield point. If the current bytecode is an OSR  yield point,
+we first store all operands that are currently on stack into pending push temporaries. An **OSR catch block** for the current inlined site index is created if it does not exist yet. An exception edge is then added between the current block and the OSR catch block.
+
+Each OSR catch block has a single successor that is called  **OSR code block**. A **prepareForOSRHelper** call node is added into that OSR code block. The node takes symbol references to all autos, parameters, and pending push temporaries used in the current method (but not methods inlined into it) as its arguments so that data flow analysis is aware  of their uses in the catch block and the subsequent optimizations will keep those symbols live. It also takes its inlined site index as an argument.
```

Comment:
takes *symbol references to* all autos → takes *loads of* all autos

Reply:
fixed